gordo_components/builder/azure_utils.py

```diff
@@ -0,0 +1,192 @@
+    """
+    Set remote tracking URI for mlflow to AzureML workspace
+
+    Parameters
+    ----------
+    azure_config: dict
+        AzureML Workspace configuration to use for remote MLFlow tracking. Passing only
+        Workspace parameters with a missing our empty `auth` key will result in
+        interactive authentication to AzureML.
+
+        Example::
+
+            `{
+                 "subscription_id":<value>,
+                 "resource_group":<value>,
+                 "workspace_name":<value>
+                 "auth": {
+                     "tenant_id":<value>,
+                     "service_principal_id":<value>,
+                     "service_principal_password":<value>
+                        }
+             }`
+
+    Returns
+    -------
+    client: mlflow.tracking.MlflowClient
+        Client with tracking uri set to AzureML if configured.
+    """
+    logger.info("Creating MLflow tracking client.")
+
+    # Get AzureML tracking_uri if using Azure as backend
+    tracking_uri = None
+    if azureml_config:
+        required = ["subscription_id", "resource_group", "workspace_name"]
+        if any([key not in azureml_config for key in required]):
+            raise ValueError(
+                f"`azureml_config` must contain keys {', '.join(required)}."
+            )
+
+        msg = "Configuring AzureML backend with {auth_type} authentication."
+        if "auth" not in azureml_config:
+            logger.info(msg.format(auth_type="interactive"))
+            azureml_config["auth"] = InteractiveLoginAuthentication(force=True)
+        else:
+            logger.info(msg.format(auth_type="ServicePrincipalAuthentication"))
+            azureml_config["auth"] = ServicePrincipalAuthentication(
+                **azureml_config["auth"]
+            )
+        tracking_uri = Workspace(**azureml_config).get_mlflow_tracking_uri()
+
+    return MlflowClient(tracking_uri)
+
+
+def get_run_id(client: MlflowClient, experiment_name: str, model_key: str) -> str:
+    """
+    Get an existing or create a new run for the given model_key and experiment_name.
+
+    The model key corresponds to a unique configuration of the model. The corresponding
+    run must be manually stopped using the `mlflow.tracking.MlflowClient.set_terminated`
+    method.
+
+    Parameters
+    ----------
+    client: mlflow.tracking.MlflowClient
+        Client with tracking uri set to AzureML if configured.
+    experiment_name: str
+        Name of experiment to log to.
+    model_key: str
+        Unique ID of model configuration.
+
+    Returns
+    -------
+    run_id: str
+        Unique ID of MLflow run to log to.
+    """
+    experiment = client.get_experiment_by_name(experiment_name)
+
+    experiment_id = (
+        getattr(experiment, "experiment_id")
+        if experiment
+        else client.create_experiment(experiment_name)
+    )
+
+    return client.create_run(experiment_id, tags={"model_key": model_key}).info.run_id
```

Comment:
Do you know what the difference between a "tag" and the metrics/params is?

Reply:
These tags are for adding some metadata to the `mlflow` `Run`. Using perhaps doesn't make as much sense as logging it as a "param"; though, this prevents the model_key from needing to be creating again inside `build_model` and added to the metadata there.